WalletWasabi.Fluent/ViewModels/Wallets/Send/TransactionPreviewViewModel.cs

```diff
@@ -22,75 +27,255 @@ public partial class TransactionPreviewViewModel : RoutableViewModel
+				{
+					UpdateTransaction(optimisePrivacyDialog.Result);
+				}
+			});
 		}
 
 		public bool PreferPsbtWorkflow => _wallet.KeyManager.PreferPsbtWorkflow;
 
-		public string AmountText { get; }
-
 		public string AddressText { get; }
 
-		public string FeeText { get; }
-
 		public string? PayJoinUrl { get; }
 
 		public bool IsPayJoin { get; }
 
+		public ICommand AdjustFeeCommand { get; }
+
+		public ICommand AvoidChangeCommand { get; }
+
+		private async Task InitialiseTransactionAsync()
+		{
+			var transactionInfo = _info;
+			var targetAnonymitySet = _wallet.ServiceConfiguration.GetMixUntilAnonymitySetValue();
+			var mixedCoins = _wallet.Coins.Where(x => x.HdPubKey.AnonymitySet >= targetAnonymitySet).ToList();
+			var totalMixedCoinsAmount = Money.FromUnit(mixedCoins.Sum(coin => coin.Amount), MoneyUnit.Satoshi);
+
+			transactionInfo.Coins = mixedCoins;
+
+			var feeDialogResult = await NavigateDialogAsync(new SendFeeViewModel(_wallet, transactionInfo, true));
+
+			if (feeDialogResult.Kind == DialogResultKind.Normal)
+			{
+				transactionInfo.FeeRate = feeDialogResult.Result;
+			}
+
+			if (transactionInfo.Amount > totalMixedCoinsAmount)
+			{
+				var privacyControlDialogResult =
+					await NavigateDialogAsync(new PrivacyControlViewModel(_wallet, transactionInfo));
+
+				if (privacyControlDialogResult.Kind == DialogResultKind.Normal &&
+				    privacyControlDialogResult.Result is { })
+				{
+					transactionInfo.Coins = privacyControlDialogResult.Result;
+				}
+			}
+		}
+
+		private async Task<BuildTransactionResult?> BuildTransactionAsync()
+		{
+			IsBusy = true;
+
+			var transactionInfo = _info;
+
+			try
+			{
+				return await Task.Run(() => TransactionHelpers.BuildTransaction(_wallet, transactionInfo));
+			}
+			catch (InsufficientBalanceException)
+			{
+				return await HandleInsufficientBalanceAsync(_wallet, transactionInfo);
+			}
+			catch (Exception ex)
+			{
+				Logger.LogError(ex);
+
+				await ShowErrorAsync("Transaction Building", ex.ToUserFriendlyString(),
+					"Wasabi was unable to create your transaction.");
+
+				return null;
+			}
+			finally
+			{
+				IsBusy = false;
+			}
+		}
+
+		private async Task<BuildTransactionResult?> HandleInsufficientBalanceAsync(Wallet wallet, TransactionInfo transactionInfo)
+		{
+			if (transactionInfo.IsPayJoin)
+			{
+				if (wallet.Coins.Sum(x => x.Amount) > transactionInfo.Amount)
+				{
+					await ShowErrorAsync("Transaction Building",
+						$"There are not enough {(transactionInfo.IsPrivatePocketUsed ? "private funds" : "funds selected")} to cover the transaction fee",
+						"Wasabi was unable to create your transaction.");
+
+					var feeDialogResult = await NavigateDialogAsync(new SendFeeViewModel(wallet, transactionInfo, false));
+					if (feeDialogResult.Kind == DialogResultKind.Normal)
+					{
+						transactionInfo.FeeRate = feeDialogResult.Result;
+					}
+
```

Comment:
// TODO ? after reducing fee.. might be enough? shouldnt we try and build txn here now? @soosr

Reply:
Yes, you are right. I am on it.